### PR TITLE
IO Serializers implementation

### PR DESCRIPTION
**Pull request by benharsh**
This PR provides a partially-stable implementation of Serializers and Deserializers, which replace the old "Formatter" prototype. This new implementation can be made the default for a program if the user compiles with the new ``-suseIOSerializers`` config param.

Otherwise, users can opt-in to using Serializers and Deserializers for individual ``fileReader`` and ``fileWriter`` instances. This is made possible by the compiler generating implementations for ``serialize`` and ``deserialize`` methods by default, as well as an initializer that implements deserialization. Compiler-generation of these methods can be disabled by compiling with ``--no-io-gen-serialization``.

Furthermore, these default implementations will, by default, try to invoke existing ``readThis`` and ``writeThis`` calls instead of the default implementation of (de)serialization. For users wishing to exclusively rely on (De)Serializers, they can compile with the flags ``--no-io-serialize-writeThis`` and ``--no-io-deserialize-readThis``.

-----

The ``fileReader`` and ``fileWriter`` types both gain new fields, named ``deserializerType`` and ``serializerType`` respectively.

These types also gain some new methods:

```chpl
// return serializer instance
proc fileWriter.serializer ref : this.serializerType

 // aliases with new serializer type
proc fileWriter.withSerializer(in serializer: ?st) : fileWriter(this.kind, this.locking, st)
proc fileWriter.withSerializer(type serializerType) : fileWriter(this.kind, this.locking, serializerType)

// return deserializer instance
proc fileReader.deserializer ref : this.deserializerType

// aliases with new deserializer type
proc fileReader.withDeserializer(in deserializer: ?dt) : fileReader(this.kind, this.locking, dt)
proc fileReader.withDeserializer(type deserializerType) : fileReader(this.kind, this.locking, deserializerType)
```

In addition, the following methods on ``file`` each gain new ``serializer`` and ``deserializer`` optional arguments with the ``in`` intent:
- openReader, openWriter
- file.reader, file.writer

-----

When ``-suseIOSerializers`` is thrown, ``fileReader`` and ``fileWriter`` will use the new ``DefaultDeserializer`` and ``DefaultSerializer`` types by default. These types implement a partially-stable API which will be detailed in a forthcoming technote. These types may attempt to invoke the following methods on user-defined classes and records, to provide users the opportunity to customize (de)serialization of their types.

```chpl
proc MyType.serialize(writer: fileWriter(?), ref serializer : writer.serializerType) throws

// For reading into existing values
proc MyType.deserialize(reader: fileReader(?), ref deserializer: reader.deserializerType) throws

// For reading in types
proc MyType.init(reader: fileReader(?), ref deserializer: reader.deserializerType) throws
```

Note that the API defines the names of these arguments, and that (De)Serializers may invoke these methods with named arguments (e.g. ``val.serialize(writer=..., serializer=...)``).

The ``serializer`` and ``deserializer`` arguments are mutable so that internal state may be modified when invoking methods on these arguments. In this initial PR such methods are unstable.

-----

Three new (De)Serializer implementations are added as package modules to support historical formats:
- Json (%jt)
- ChplFormat (%ht)
- BinaryIO (iokind, and other options)

Now, these formats are decoupled from the IO infrastructure and can be improved without modifying the runtime. New formats can also be added much more easily than in the past.

-----

For the time being, stable code can use the aforementioned methods on ``fileReader`` and ``fileWriter`` with DefaultSerializer/Deserializer and with the compiler-generated default implementations of ``serialize`` and ``deserialize`` methods. The methods that will work through (De)Serializers are currently unstable.

-----

Testing:
- [x] full local
- [x] local with --no-io-gen-serialization
- [x] local with -suseIOSerializers (will still have errors until we can change .good files)
- [x] full gasnet
- [x] memLeaks

### CODE REVIEW

File: test/library/draft/DataFrames/DataFrames.chpl [hunk 1]
```diff
@@ -65,7 +65,7 @@ module DataFrames {
     }
 
     @chpldoc.nodoc
-    proc writeThis(f) throws {
+    override proc writeThis(f) throws {
```
Comment (jeremiah-corrado) [inline test/library/draft/DataFrames/DataFrames.chpl:68]: Same question here about the `override`...

**Review by jeremiah-corrado — COMMENTED** [review]
Overall looks great! I didn't have a chance to look at the compiler changes yet, but I plan to do that tomorrow.

A few meta notes: 
- The new package modules are not showing up in the docs yet. (I believe they will need to be added to `doc/rst/meta/modules` and the PACKAGES_TO_DOCUMENT section of the modules makefile)
- I noticed some code changes to bulk array IO, particularly in DefaultRectangular.chpl. Are those changes related to future work with more efficient array serialization, or are they already tested in io/serializers/arrays?
- Out of curiosity: many of the serialize/deserialize methods on types in the standard library are implemented in terms of their writeThis/readThis. Do those get replaced by compiler generated implementations, or are those just cases where readThis/writeThis will do the right thing? 

I also have a few documentation suggestions and questions:

File: modules/internal/DefaultAssociative.chpl [hunk 2]
```diff
@@ -641,6 +641,71 @@ module DefaultAssociative {
...
+      fmt.endMap(f);
+    }
+
+    proc _isDefaultDeser(f) param : bool {
+      if f.writing then return f.serializerType == IO.DefaultSerializer;
+      else return f.deserializerType == IO.DefaultDeserializer;
+    }
+
+    proc dsiSerialReadWrite(f, in printBraces=true, inout first = true) throws
+    where _isDefaultDeser(f) {
```
[the remark this answers is not in the document]
Comment (benharsh) [reply modules/internal/DefaultAssociative.chpl:684]: ``_isDefaultDeser(f)`` returning ``true`` implies that it has a serializer. ``_usingSerializers`` checks whether the (de)serializerType is ``nothing``, which won't be true if we're actually using a Default(De)Serializer.

File: modules/packages/LinkedLists.chpl [hunk 3]
```diff
@@ -376,11 +405,53 @@ record LinkedList {
       if isJson || isChpl then f._readLiteral("]");
   }
 
+  proc deserialize(reader: fileReader, ref deserializer) throws
+  where reader.deserializerType == IO.DefaultDeserializer {
```
[the remark this answers is not in the document]
Comment (benharsh) [reply modules/packages/LinkedLists.chpl:409]: I just did this to match the current behavior on main for our "default" format. I think the ``List`` format makes more sense for an unknown serialization format.

File: modules/packages/TOML.chpl [hunk 4]
```diff
@@ -1358,12 +1358,13 @@ module TomlReader {
...
 
-    proc init(r: fileReader) {
+    @chpldoc.nodoc
+    proc init(reader: fileReader, ref deserializer) {
       this.complete();
       compilerError("Reading a Tokens type is not supported");
     }
 
-    proc writeThis(f) throws {
+    override proc writeThis(f) throws {
```
[the remark this answers is not in the document]
Comment (benharsh) [reply modules/packages/TOML.chpl:1367]: Technically everything inherits from the root ``object``, and so we _should_ add ``override`` for correctness. I don't know that it's actually necessary, truth be told.

File: modules/standard/IO.chpl [hunk 5]
```diff
@@ -2470,15 +2481,15 @@ record fileReader {
...
+  type deserializerType = defaultSerializeType(/* writing= */ false);
 
   @chpldoc.nodoc
   var _home:locale = here;
   @chpldoc.nodoc
   var _channel_internal:qio_channel_ptr_t = QIO_CHANNEL_PTR_NULL;
 
   @chpldoc.nodoc
-  var _fmt : fmtType;
+  var _deserializer : unmanaged _serializeWrapper?(deserializerType);
```
[the remark this answers is not in the document]
Comment (benharsh) [reply modules/standard/IO.chpl:2492]: I've done a couple of spot checks on tests in the past, but I'll go ahead and do a large run just to be safe.

**Comment by benharsh** [comment]
> I noticed some code changes to bulk array IO, particularly in DefaultRectangular.chpl. Are those changes related to future work with more efficient array serialization, or are they already tested in io/serializers/arrays?

BinaryIO implements the optional API, and so any tests in ``io/serializers/`` that work with DefaultRectangular and BinaryIO will exercise that capability.

> Out of curiosity: many of the serialize/deserialize methods on types in the standard library are implemented in terms of their writeThis/readThis. Do those get replaced by compiler generated implementations, or are those just cases where readThis/writeThis will do the right thing?

I believe these are cases where the ``readThis`` or ``writeThis`` will do the right thing. Either that right thing is always printing out a type in a particular way, or perhaps just invoking a compilerError.

**Comment by mppf** [comment]
I've reviewed the compiler changes (only).

**Review by jeremiah-corrado — APPROVED** [review]

**Pull request merged**